### PR TITLE
build proxy with conditional logging and jsonselect encoders

### LINKED ISSUE #4740: [proxy] Produce access logs only for non-200 requests
Thereby cleaning up the proxy logs and making them more useful.

### PR DESCRIPTION
**Pull request by leodido**
Fixes #4740 (do not log 200 requests).

Also:
- do not log when status code is 0 or 304 too
- reshape the resulting JSON log as per Stackdriver Log Entry format.

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4864#" title="Author self-approved">leodido</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **jankeromnes** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @jankeromnes` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by leodido** [comment]
Reminder for myself tomorrow: debug the `"status":0` logs.

Aside from those, it is working - ie., not emitting all the non 200 access logs.

image

Also [stackdriver log entry](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry) format looks rad.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4864#" title="Author self-approved">leodido</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **svenefftinge** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @svenefftinge` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by aledbf** [comment]
@leodido please also exclude the status code 304

**Comment by aledbf** [comment]
@leodido I just tried to run some of the examples and I see some 200
`{"severity":"INFO","timestamp":"2021-07-20T11:23:42.502300441Z","logName":"http.log.access.log2","httpRequest":{"requestMethod":"PUT","protocol":"HTTP/1.1","status":200,"responseSize":0,"userAgent":"MinIO (linux; amd64) minio-go/v7.0.11"}}
`

Also, I think we should add the path of the request.

**Comment by leodido** [comment]
Yes Alex you're right. I found the same too.

There's a bug in the expression evaluator I'm fixing

**Comment by leodido** [comment]
Ok, these are some logs now:


```json
{"severity":"INFO","timestamp":"2021-07-21T09:39:20.042817513Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":302,"responseSize":0,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/api/login?host=github.com&returnTo=https%3A%2F%2Fleo-caddy-logger-plugin.staging.gitpod-dev.com%2Fcomplete-auth%3Fmessage%3Dsuccess"}}
{"severity":"INFO","timestamp":"2021-07-21T09:39:21.051060152Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":302,"responseSize":190,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/auth/github/callback?code=b905eb14ab149f4a97e1&state=leo-caddy-logger-plugin"}}
{"severity":"INFO","timestamp":"2021-07-21T09:39:21.264285932Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":302,"responseSize":212,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/api/tos?mode=login"}}
{"severity":"ERROR","timestamp":"2021-07-21T09:41:06.096766455Z","logName":"http.log.access.log2","httpRequest":{"requestMethod":"GET","protocol":"HTTP/1.1","status":404,"responseSize":369,"userAgent":"MinIO (linux; amd64) minio-go/v7.0.11","requestUrl":"/gitpod-user-7ba328d8-0454-4a26-b4d0-1adb8badefa2/?location="}}
{"severity":"ERROR","timestamp":"2021-07-21T09:41:06.189535151Z","logName":"http.log.access.log2","httpRequest":{"requestMethod":"HEAD","protocol":"HTTP/1.1","status":404,"responseSize":0,"userAgent":"MinIO (linux; amd64) minio-go/v7.0.11","requestUrl":"/gitpod-user-7ba328d8-0454-4a26-b4d0-1adb8badefa2/workspaces/violet-ermine-wwnq93qg/full.tar"}}
{"severity":"INFO","timestamp":"2021-07-21T09:41:06.758797925Z","logName":"http.log.access.log1","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":303,"responseSize":193,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/favicon.ico"}}
{"severity":"INFO","timestamp":"2021-07-21T09:42:18.086951018Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":204,"responseSize":0,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/code-sync/v1/resource/globalState/latest"}}
{"severity":"INFO","timestamp":"2021-07-21T09:42:20.573728889Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":204,"responseSize":0,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/code-sync/v1/resource/keybindings/latest"}}
{"severity":"INFO","timestamp":"2021-07-21T09:42:20.574281602Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":204,"responseSize":0,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/code-sync/v1/resource/snippets/latest"}}
{"severity":"INFO","timestamp":"2021-07-21T09:42:29.531150562Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":204,"responseSize":0,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/code-sync/v1/resource/machines/latest"}}
{"severity":"INFO","timestamp":"2021-07-21T09:42:29.710513684Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":204,"responseSize":0,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/code-sync/v1/resource/keybindings/latest"}}
{"severity":"INFO","timestamp":"2021-07-21T09:42:29.942736795Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":204,"responseSize":0,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/code-sync/v1/resource/snippets/latest"}}
{"severity":"INFO","timestamp":"2021-07-21T09:42:30.050469218Z","logName":"http.log.access.log0","httpRequest":{"requestMethod":"GET","protocol":"HTTP/2.0","status":204,"responseSize":0,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0","requestUrl":"/code-sync/v1/resource/globalState/latest"}}
```

I think this is ready for review now. :)

File: components/proxy/conf/Caddyfile [hunk 1]
```diff
@@ -50,16 +50,9 @@
...
-			wrap json
-			fields {
-				logger			delete
-				msg				delete
-				size			delete
-				status			delete
-				resp_headers	delete
-				request			delete
-			}
+		format if "status != 200 && status != 0 && status != 304" jsonselect "{severity:level} {timestamp:ts} {logName:logger} {httpRequest>requestMethod:request>method} {httpRequest>protocol:request>proto} {httpRequest>status:status} {httpRequest>responseSize:size} {httpRequest>userAgent:request>headers>User-Agent>[0]} {httpRequest>requestUrl:request>uri}" {
```
Comment (aledbf) [inline components/proxy/conf/Caddyfile:53]: @leodido maybe we can just use `status > 399`?

**Comment by aledbf** [comment]
Besides my comment to simplify the filter this change LGTM

**Comment by aledbf** [comment]
@leodido thank you for working on this!!!

File: components/proxy/conf/Caddyfile (hunk 1 above)
Comment (leodido) [reply components/proxy/conf/Caddyfile:53]: if that's what we want I can implement the `>` operator and do it this way, ofc :)

Comment (aledbf) [reply components/proxy/conf/Caddyfile:53]: @leodido let's do that in a future PR :)

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 44b9fbce388ab77b483eb7d220111b868e07cead</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4864#issuecomment-884132466" title="LGTM">aledbf</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4864#" title="Author self-approved">leodido</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **svenefftinge** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @svenefftinge` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4864#issuecomment-884133155" title="Approved">aledbf</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4864#" title="Author self-approved">leodido</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **svenefftinge** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @svenefftinge` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/proxy/Dockerfile [hunk 2]
```diff
@@ -25,6 +27,8 @@ RUN git clone -b http2 https://github.com/aledbf/caddy caddy-fork
 RUN xcaddy build \
   --output /caddy \
   --with github.com/caddyserver/caddy/v2=$PWD/caddy-fork \
+  --with github.com/leodido/caddy-conditional-logging \
+  --with github.com/leodido/caddy-jsonselect-encoder \
```
Comment (csweichel) [inline components/proxy/Dockerfile:31]: Please pull the plugins into this repo. That would also solve the leeway rebuilding issue.

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 0fbec4a928fd0db5b5a94d1f6f01ce2456a07c99</details>

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

**Comment by leodido** [comment]
Re-organized the caddy plugins into separated go modules.

This is the current custom caddy plugin list:

```console
caddy.logging.encoders.if
caddy.logging.encoders.jsonselect
http.handlers.gitpod_cors_origin
http.handlers.gitpod_sec_websocket_key
http.handlers.gitpod_workspace_download

  Non-standard modules: 5
```

**Review by csweichel — COMMENTED** [review]
other than the nit, this looks good to me

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 5b043e37e7aa5b126e6453483aa8d4eba0eec11e</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4864#issuecomment-884954894" title="LGTM">aledbf</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4864#issuecomment-886739232" title="LGTM">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4864#" title="Author self-approved">leodido</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **svenefftinge** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @svenefftinge` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4864#issuecomment-886746998" title="Approved">aledbf</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4864#issuecomment-886739232" title="LGTM">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4864#" title="Author self-approved">leodido</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **svenefftinge** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @svenefftinge` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by leodido** [comment]
Ok @roboquat keep calm ahaha!

I've asked people in the @gitpod-io/engineering-meta  to take a look at this 👼

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4864#issuecomment-886746998" title="Approved">aledbf</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4864#issuecomment-886848743" title="Approved">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4864#" title="Author self-approved">leodido</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**